File: 04_arrays_part4/6_longest_substring_without_repeating_character.go
func lengthOfLongestSubstring(s string) int {
	n := len(s)
	if n == 0 {
		return 0
	}
	left, res, seen := 0, 0, map[rune]int{}
	for i, c := range s {
		if val, ok := seen[c]; ok && val >= left {
			res = max(res, i-left)
			left = val + 1
		}
		seen[c] = i
	}
	return max(res, n-left)
}

File: 04_arrays_part4/6_longest_substring_without_repeating_character.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        if n == 0:
            return 0
        left = res = 0
        seen = {}
        for i in range(n):
            if s[i] in seen and seen[s[i]] >= left:
                res = max(res, i - left)
                left = seen[s[i]] + 1
            seen[s[i]] = i

        return max(res, i - left + 1)
